File: src/bindings/nodejs/typescript_oxlint/ts/rules/no_unsafe_unary_minus.ts
import { createNativeRule } from "./rule_creator";
import { classifyTypeText, splitTypeText, typeTextsAtNode } from "./type_utils";

export const noUnsafeUnaryMinusRule = createNativeRule(
  "no-unsafe-unary-minus",
  {
    docs: {
      description: "Disallow unary negation on non-number and non-bigint values.",
    },
    messages: {
      unaryMinus: "Argument of unary negation should be assignable to number | bigint.",
    },
  },
  (context) => ({
    UnaryExpression(node: any) {
      if (node.operator !== "-") {
        return;
      }
      if (isSafeLiteral(node.argument)) {
        return;
      }
      const typeTexts = typeTextsAtNode(context, node.argument);
      if (
        typeTexts.length > 0 &&
        typeTexts.every((text) => {
          return splitTypeText(text).every((part) => {
            const kind = classifyTypeText(part);
            return kind === "any" || kind === "number" || kind === "bigint";
          });
        })
      ) {
        return;
      }
      context.report({ node, messageId: "unaryMinus" });
    },
  }),
);

function isSafeLiteral(node: any): boolean {
  return (
    node?.type === "Literal" && (typeof node.value === "number" || typeof node.bigint === "string")
  );
}

File: src/core/corsa_core/src/lint/rules/no_unsafe_unary_minus.rs
use super::super::{LintNode, RuleContext, RuleMessage, RustLintRule};
use crate::lint::helpers::{is_number_or_bigint_literal, is_unary_minus_type_safe};

/// Type-aware rule that rejects unary minus on non-number-like values.
#[derive(Clone, Copy, Debug, Default)]
pub struct NoUnsafeUnaryMinusRule;

const MESSAGES: &[RuleMessage] = &[RuleMessage {
    id: "unaryMinus",
    description: "Argument of unary negation should be assignable to number | bigint.",
}];
const LISTENERS: &[&str] = &["UnaryExpression"];

impl RustLintRule for NoUnsafeUnaryMinusRule {
    fn name(&self) -> &'static str {
        "no-unsafe-unary-minus"
    }

    fn docs_description(&self) -> &'static str {
        "Disallow unary negation on non-number and non-bigint values."
    }

    fn messages(&self) -> &'static [RuleMessage] {
        MESSAGES
    }

    fn listeners(&self) -> &'static [&'static str] {
        LISTENERS
    }

    fn check(&self, ctx: &mut RuleContext<'_>, node: &LintNode) {
        if node.kind != "UnaryExpression" || node.field_str("operator") != Some("-") {
            return;
        }
        let Some(argument) = node.child("argument") else {
            return;
        };
        if is_number_or_bigint_literal(argument) || is_unary_minus_type_safe(&argument.type_texts) {
            return;
        }
        ctx.report("unaryMinus", node.range);
    }
}
